shim/shim.py

```diff
@@ -612,147 +612,148 @@ def _add_debug_arguments(parser: argparse.ArgumentParser) -> None:
-        options.append(arguments.sharded_models_directory)
+        options_json["sharded-models-directory"] = arguments.sharded_models_directory
 
     if arguments.emit_all_via_cast_features:
-        options.append("--emit-all-via-cast-features")
+        options_json["emit-all-via-cast-features"] = arguments.emit_all_via_cast_features
+
     if arguments.propagate_across_arguments:
-        options.append("--propagate-across-arguments")
+        options_json["propagate-across-arguments"] = arguments.propagate_across_arguments
+
     if arguments.allow_via_cast_feature:
+        options_json["allow-via-cast-feature"] = []
         for feature in arguments.allow_via_cast_feature:
-            options.append("--allow-via-cast-feature=%s" % feature.strip())
+            options_json["allow-via-cast-feature"].append(feature.strip())
 
     if arguments.heuristics:
-        options.append("--heuristics")
-        options.append(arguments.heuristics)
+        options_json["heuristics"] = arguments.heuristics
 
     if arguments.sequential:
-        options.append("--sequential")
+        options_json["sequential"] = arguments.sequential
+
     if arguments.skip_source_indexing:
-        options.append("--skip-source-indexing")
+        options_json["skip-source-indexing"] = arguments.skip_source_indexing
+
     if arguments.skip_analysis:
-        options.append("--skip-analysis")
+        options_json["skip-analysis"] = arguments.skip_analysis
+
     if arguments.disable_parameter_type_overrides:
-        options.append("--disable-parameter-type-overrides")
+        options_json["disable-parameter-type-overrides"] = arguments.disable_parameter_type_overrides
+
     if arguments.disable_global_type_analysis:
-        options.append("--disable-global-type-analysis")
+        options_json["disable-global-type-analysis"] = arguments.disable_global_type_analysis
+
     if arguments.verify_expected_output:
-        options.append("--verify-expected-output")
+        options_json["verify-expected-output"] = arguments.verify_expected_output
+
     if arguments.remove_unreachable_code:
-        options.append("--remove-unreachable-code")
+        options_json["remove-unreachable-code"] = arguments.remove_unreachable_code
+
     if arguments.maximum_method_analysis_time is not None:
-        options.append("--maximum-method-analysis-time")
-        options.append(str(arguments.maximum_method_analysis_time))
+        options_json["maximum-method-analysis-time"] = arguments.maximum_method_analysis_time
+
     if arguments.enable_cross_component_analysis:
-        options.append("--enable-cross-component-analysis")
+        options_json["enable-cross-component-analysis"] = arguments.enable_cross_component_analysis
+
     if arguments.extra_analysis_arguments:
-        options.extend(shlex.split(arguments.extra_analysis_arguments))
+        for arg in shlex.split(arguments.extra_analysis_arguments):
+            options_json[arg] = arg
```

Comment:
arg, I forgot about this. We used this to inject extra parameters for the C++ binary.
I don't think your implementation works, for instance we would be calling it like `--extra-analysis-arguments=--enable-cross-component-analysis`.
I propose to let users pass arbitrary json and merge it with the options. Something like:
```
if arguments.extra_analysis_arguments:
   extra_arguments = json.loads(arguments.extra_analysis_arguments)
   for key, value in extra_arguments.items():
     if key in options_json and isinstance(options_json[key], list) and isinstance(value, list):
       # append to the existing options
       options_json[key].extend(value)
     else:
      # override
      options_json[key] = value
```
This should also be moved towards the end of this function